### PR TITLE
Permission-gate fields when content is deleted

### PR DESCRIPTION
**Pull request by oetherington**
Currently, the body of deleted content is available in the Apollo cache that's embedded in the HTML during SSR. This is a long-standing problem, as is evidenced by Jim's comment in the code from mid-2021.

This PR adds a new permissions function to disallow this, but we still make the content available to admins, mods and the original author.

By default, I've applied this to all fields that use `make_editable`, but it's plausible that there's some reason I've not thought about that means we should restrict to to fewer fields, or apply it to more.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1204535640179844) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/lib/collections/comments/collection.ts [hunk 1]
```diff
@@ -53,27 +53,6 @@ export const Comments: ExtendedCommentsCollection = createCollection({
   logChanges: true,
 });
 
-Comments.checkAccess = async (currentUser: DbUser|null, comment: DbComment, context: ResolverContext|null): Promise<boolean> => {
-  if (userIsAdmin(currentUser) || userOwns(currentUser, comment)) { // admins can always see everything, users can always see their own posts
-    return true;
-  } else {
-    return true;
-  }
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/comments/collection.ts:61]: the fuck was this

File: packages/lesswrong/lib/collections/comments/collection.ts [hunk 2]
```diff
@@ -53,27 +53,6 @@ export const Comments: ExtendedCommentsCollection = createCollection({
...
-  
-  // NOTE: There used to be a special case here that would deny access (to non-
-  // admins) if comment.isDeleted was true. This was wrong in two cancelling
-  // ways: first, because we show UI on post pages indicating that a deleted
-  // comment used to be there, and denying access would hide that UI. And
-  // second, because the field is named `deleted`, not named `isDeleted`.
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/comments/collection.ts:67]: what a terrible codebase we used to have

File: packages/lesswrong/lib/vulcan-users/permissions.ts [hunk 3]
```diff
@@ -114,6 +116,26 @@ export const userOwns = function (user: UsersMinimumInfo|DbUser|null, document:
...
+  // Admins and mods can see deleted content
+  if (userIsAdminOrMod(user)) {
+    return true;
+  }
+  // Authors can see their deleted content
+  if (userOwns(user, document)) {
+    return true;
+  }
+  // Unfortunately, different collections use different field names
+  // to represent "deleted-ness"
```
Comment (jpaddison3) [inline packages/lesswrong/lib/vulcan-users/permissions.ts:132]: Very unfortunate indeed

File: packages/lesswrong/lib/vulcan-users/permissions.ts [hunk 4]
```diff
@@ -114,6 +116,26 @@ export const userOwns = function (user: UsersMinimumInfo|DbUser|null, document:
   }
 };
 
+export const documentIsNotDeleted = (
```
Comment (jpaddison3) [inline packages/lesswrong/lib/vulcan-users/permissions.ts:119]: Name is maybe a bit misleading

**Review by jpaddison3 — CHANGES_REQUESTED** [review]
I basically like this, but I would like to see access to the author field(s) to also be removed, for: comments and posts.

File: packages/lesswrong/lib/vulcan-users/permissions.ts (hunk 4 above)
Comment (oetherington) [reply packages/lesswrong/lib/vulcan-users/permissions.ts:119]: Yeah, I wondered about this but couldn't come up with anything better. It seems fine to me in the context of being used alongside `userOwns`, along with it being in `permissions.ts`. Suggestions are welcome though.

**Comment by oetherington** [comment]
> I basically like this, but I would like to see access to the author field(s) to also be removed, for: comments and posts.

Done!

File: packages/lesswrong/lib/collections/comments/schema.ts [hunk 5]
```diff
@@ -413,7 +413,7 @@ const schema: SchemaType<DbComment> = {
   // DEPRECATED field for GreaterWrong backwards compatibility
   htmlBody: resolverOnlyField({
     type: String,
-    canRead: ['guests'],
+    canRead: [documentIsNotDeleted],
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/comments/schema.ts:416]: nice

File: packages/lesswrong/lib/collections/posts/schema.tsx [hunk 6]
```diff
@@ -2157,7 +2157,7 @@ const schema: SchemaType<DbPost> = {
   shareWithUsers: {
     type: Array,
     order: 15,
-    canRead: ['guests'],
+    canRead: [documentIsNotDeleted],
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/posts/schema.tsx:2160]: Wait is this even the right thing? That seems too permissive.

File: packages/lesswrong/lib/collections/posts/schema.tsx
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/posts/schema.tsx:1]: Going wild in this file and I love it

**Review by jpaddison3 — APPROVED** [review]
Great updates.

File: packages/lesswrong/lib/collections/posts/schema.tsx (hunk 6 above)
Comment (oetherington) [reply packages/lesswrong/lib/collections/posts/schema.tsx:2160]: Without actually checking, I'd that if you make the permissions tighter here then maybe shared users can't tell that they're shared(?).

If that's not the case then it's probably a problem for another PR.

**Pull request merged**